src/mame/drivers/uknc.cpp

```diff
@@ -9,97 +9,946 @@ Graphics 640x288 pixels.
+			vdata |= (BIT(data, 4) ? BIT(m_video.foreground, 2) : BIT(m_video.background, 18)) << 4;
+			vdata |= (BIT(data, 5) ? BIT(m_video.foreground, 2) : BIT(m_video.background, 22)) << 5;
+			vdata |= (BIT(data, 6) ? BIT(m_video.foreground, 2) : BIT(m_video.background, 26)) << 6;
+			vdata |= (BIT(data, 7) ? BIT(m_video.foreground, 2) : BIT(m_video.background, 30)) << 7;
+			m_videoram[m_video.vram_addr + 0100000] = vdata;
+		}
+		break;
+
+	case 4: // 177026
+		m_video.plane_mask = data & 7;
+		break;
+	}
+}
+
+// system registers
+
+/*
+ *	00	R	cassette data in
+ *	01	RW	cassette data out
+ *	02	RW	cassette data gate
+ *	03	RW	cassette/external_event/index gate.  0: cassette  1: ???
+ *	04	RW	maincpu USER/HALT mode. 0: USER, 1: HALT
+ *	05	RW	maincpu DCLO pin
+ *	06	-
+ *	07	RW	sound data out, if 08-12 == 0
+ *	12:08	sound frequency matrix
+ *	13	RW	reserved
+ *	14	-
+ *	15	RW	maincpu ACLO pin
+ */
+uint16_t uknc_state::scsr_r()
+{
+	double tap_val = m_cassette->input();
+
+	m_scsr = (m_scsr & ~1) | (tap_val < 0 ? 1 : 0);
+
+	return m_scsr;
+}
+
+void uknc_state::scsr_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	m_scsr = data;
+
+	if (BIT(data, 2)) m_cassette->output(BIT(data, 1) ? -1 : 1);
+
+// requires real VM2 cpu core
+#if 0
+	m_maincpu->set_input_line(INPUT_LINE_1801HALT, BIT(data, 4) ? ASSERT_LINE : CLEAR_LINE);
+
+	if (BIT(data, 15) && m_maincpu->suspended(SUSPEND_REASON_DISABLE))
+	{
+		m_maincpu->resume(SUSPEND_REASON_DISABLE);
+	}
+	else if (!BIT(data, 15) && !m_maincpu->suspended(SUSPEND_REASON_DISABLE))
+	{
+		m_maincpu->set_input_line(INPUT_LINE_ACLO, ASSERT_LINE);
+	}
+#endif
+}
+
+/*
+ *	0-3	CE0..CE3 bus signals
+ *	  CE0	mapping at 100000: 0 - read from VRAM plane 0, 1 - read from ROM (reset to 1)
+ *	  CE1-2	ROM carts: bank selector
+ *	  CE3	slot number
+ *	4-7	ROM banks replacement: 0 - writes ignored, 1 - writes go to VRAM plane 0
+ *	  4		100000..117777
+ *	  5		120000..137777
+ *	  6		140000..157777
+ *	  7		160000..177777
+ *	8	line clock interrupt (subcpu): 0 - enable
+ *	9	line clock interrupt (maincpu): 0 - enable
+ */
+uint16_t uknc_state::bcsr_r()
+{
+	return m_bcsr;
+}
+
+void uknc_state::bcsr_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	int bank1, bank2;
+	address_space *subspace = &m_subcpu->space(AS_PROGRAM);
+
+	bank1 = (data & 1) | ((data >> 3) & 1);
+	bank2 = (data >> 5) & 7;
+
+	// order matters: ROM on device may unmap itself and we restore the mapping
+//	m_subqbus->ce0_ce3_w(data & 15);
+
+	if (BIT(m_bcsr ^ data, 0) && BIT(data, 0))
+	{
+		subspace->unmap_readwrite(0100000, 0117777);
+		subspace->install_readwrite_handler(0100000, 0117777,
+			read16s_delegate(m_bankdev1, FUNC(address_map_bank_device::read16)),
+			write16s_delegate(m_bankdev1, FUNC(address_map_bank_device::write16)));
+	}
+
+	m_bankdev1->set_bank(bank1);
+	m_bankdev2->set_bank(bank2);
+
+	m_bcsr = data;
+}
+
+// video
+
+#define YRGB(rgb, y, p) ((rgb)*42 * (3 + (p) + (y)*2))
```

Comment:
Don’t use macros unnecessarily.  You can use a local non-capturing lambda or a `constexpr` function.

Reply:
Is there an example I can use?

Reply:
Lambda syntax is where you have a local like `auto YRGB = [] (auto rgb, auto y, auto p) { return ...; };`

Or you could make it a `constexpr` function, which is a lot like any other function, but with a `constexpr` qualifier, and its body needs to be free of global references and side effects.

But there’s no reason to resort to a macro for this.